File: app/src/main/java/org/thoughtcrime/securesms/backup/BackupEvent.java
package org.thoughtcrime.securesms.backup;

public class BackupEvent {
  public enum Type {
    PROGRESS,
    PROGRESS_VERIFYING,
    FINISHED
  }

  private final Type type;
  private final long count;
  private final long estimatedTotalCount;

  public BackupEvent(Type type, long count, long estimatedTotalCount) {
    this.type                = type;
    this.count               = count;
    this.estimatedTotalCount = estimatedTotalCount;
  }

  public Type getType() {
    return type;
  }

  public long getCount() {
    return count;
  }

  public long getEstimatedTotalCount() {
    return estimatedTotalCount;
  }

  public double getCompletionPercentage() {
    if (estimatedTotalCount == 0) {
      return 0;
    }

    return Math.min(99.9f, (double) count * 100L / (double) estimatedTotalCount);
  }
}

File: messenger/src/main/java/org/thoughtcrime/securesms/backup/BackupEvent.kt
package org.thoughtcrime.securesms.backup

data class BackupEvent constructor(val type: Type, val count: Int, val exception: Exception?) {

    enum class Type {
        PROGRESS, FINISHED
    }

    companion object {
        @JvmStatic fun createProgress(count: Int) = BackupEvent(Type.PROGRESS, count, null)
        @JvmStatic fun createFinished() = BackupEvent(Type.FINISHED, 0, null)
        @JvmStatic fun createFinished(e: Exception?) = BackupEvent(Type.FINISHED, 0, e)
    }
}